File: worker/src/core/hijri-date-phrase.mjs
/**
 * Zerlegt eine hidschri-Datumsangabe in ihre EINZELNEN belegten Möglichkeiten.
 *
 * Hintergrund: die Quellen nennen für eine Person regelmäßig mehrere Jahre
 * nebeneinander — «145، أو: 146هـ، أو: 147هـ، وقيل: 144هـ» sind vier Angaben,
 * nicht eine. Bis hierher hat der Atlas davon genau eine gezeigt
 * (`death_year_ah`), womit drei belegte Aussagen unsichtbar blieben. Das
 * widerspricht Abschnitt 7 der Projektbeschreibung: widersprüchliche
 * Datierungen müssen gleichzeitig sichtbar sein, und es wird nichts gemittelt,
 * gerundet oder zu einer Spanne verschmolzen.
 *
 * Diese Funktion rechnet deshalb NICHT. Sie liest die Jahre, die dastehen, und
 * behält den Wortlaut jeder einzelnen Angabe bei.
 *
 * Der Unterschied zwischen den Verbindungen ist fachlich, nicht kosmetisch:
 *   «أو»   — eine alternative Lesung INNERHALB derselben Aussage,
 *   «وقيل» — eine EIGENSTÄNDIGE Aussage einer anderen Autorität.
 * Beide bleiben getrennt gekennzeichnet, damit die Oberfläche sie nicht als
 * gleichrangige Zahlenreihe darstellt.
 */

const ARABIC_INDIC = "٠١٢٣٤٥٦٧٨٩";

/** Wandelt arabisch-indische Ziffern in ASCII, sonst unverändert. */
function toAsciiDigits(value) {
  let out = "";
  for (const char of value) {
    const index = ARABIC_INDIC.indexOf(char);
    out += index >= 0 ? String(index) : char;
  }
  return out;
}

/**
 * Näherungs- und Grenzangaben. Sie ändern die Bedeutung des Jahres und dürfen
 * nicht verschwinden: «مات بعد 140» nennt kein Todesjahr 140.
 */
const QUALIFIERS = [
  ["بعيد", "shortly_after"],
  ["قبيل", "shortly_before"],
  ["بعد", "after"],
  ["قبل", "before"],
  ["نحو", "circa"],
  ["حدود", "circa"],
  ["زهاء", "circa"],
];

/** «وقيل»/«قيل» führt eine eigenständige Aussage ein, «أو» eine Lesart derselben. */
function relationFor(connector, isFirst) {
  if (isFirst) return "primary";
  if (/ق(?:ي|ِي)ل/.test(connector)) return "reported";
  if (/أو|او/.test(connector)) return "alternative";
  return "additional";
}

function qualifierFor(segment) {
  for (const [word, code] of QUALIFIERS) {
    if (segment.includes(word)) return code;
  }
  return null;
}

/**
 * Liest alle Jahresangaben einer Phrase.
 *
 * @param {string|null|undefined} phrase Wortlaut aus der Quelle
 * @param {"birth"|"death"} kind
 * @returns {{assertions: Array<object>, sourcePhrase: string|null}}
 */
export function parseHijriYearPhrase(phrase, kind) {
  const sourcePhrase = typeof phrase === "string" && phrase.trim() ? phrase.trim() : null;
  if (!sourcePhrase) return { assertions: [], sourcePhrase: null };

  const ascii = toAsciiDigits(sourcePhrase);
  const assertions = [];
  const seen = new Set();
  const pattern = /(\d{1,4})/g;
  let match;
  let previousEnd = 0;

  while ((match = pattern.exec(ascii)) !== null) {
    const year = Number(match[1]);
    // Ein Jahr über 1100 AH ist in diesem Bestand keine Jahresangabe, sondern
    // eine Seiten- oder Bandzahl, die versehentlich in der Phrase steht.
    if (!year || year > 1100) { previousEnd = match.index + match[0].length; continue; }

    // Der Text zwischen der letzten Zahl und dieser trägt die Verbindung
    // («أو:», «وقيل:») und einen möglichen Grenzzusatz.
    const connector = ascii.slice(previousEnd, match.index);
    const isFirst = assertions.length === 0;
    const relation = relationFor(connector, isFirst);
    const qualifier = qualifierFor(isFirst ? ascii.slice(0, match.index) : connector);

    // Der angezeigte Wortlaut umfasst die Verbindung und das Jahr samt «هـ»,
    // damit in der Oberfläche «وقيل: 144هـ» steht und nicht bloß «144».
    const tailMatch = /^\s*(?:هـ|هجرية|هج)?/.exec(ascii.slice(match.index + match[0].length));
    const wordingStart = isFirst ? match.index : previousEnd;
    const wording = sourcePhrase.slice(wordingStart, match.index + match[0].length + (tailMatch ? tailMatch[0].length : 0))
      // Das «هـ» am Anfang gehört noch zur VORIGEN Angabe und würde sonst
      // jede Folgeangabe als «هـ، وقيل: 144هـ» ausgeben.
      .replace(/^[\s،؛,.:]*(?:هـ|هجرية|هج)?[\s،؛,.:]*/, "")
      .trim();

    const key = `${year}|${relation}|${qualifier ?? ""}`;
    if (!seen.has(key)) {
      seen.add(key);
      // Feldbild von ApiRijalDateAssertion (lib/api-client.ts) — kein zweites
      // Vokabular fuer dieselbe Sache.
      assertions.push({
        kind,
        verb: null,
        qualifier,
        valueAh: year,
        approximate: qualifier === "circa",
        rawPhrase: wording || String(year),
        textOffset: match.index,
        evidenceClass: "rijal_statement",
        confidence: null,
        reviewStatus: "machine_unreviewed",
        // «أو» (Lesart derselben Aussage) gegen «وقيل» (eigenstaendige Aussage).
        relation,
        // Kein Wert ohne Rueckfuehrbarkeit: die vollstaendige Quellphrase
        // bleibt an jeder einzelnen Angabe haengen.
        sourcePhrase,
      });
    }
    previousEnd = match.index + match[0].length;
  }

  return { assertions, sourcePhrase };
}

/**
 * Baut die Datumsangaben eines Rijāl-Eintrags aus den beiden Wortlautspalten.
 *
 * Bewusst OHNE Rückfall auf `death_year_ah`/`birth_year_ah`: diese Liste sagt
 * aus, was im Wortlaut der Quelle steht. Nennt die Quelle kein Jahr in Ziffern
 * — etwa in den Turath-Werken, die es ausgeschrieben führen («ست وثلاثين
 * ومئتين») — bleibt die Liste leer, statt eine Aussage zu erzeugen, für die es
 * hier keinen Beleg gibt. Das ausgewertete Einzeljahr steht weiterhin in
 * `deathYearCandidate`.
 *
 * @param {Record<string, any>} row Zeile aus rijal_entry_ref
 */
export function rijalDateAssertions(row) {
  return [
    ...parseHijriYearPhrase(row.birth_original_phrase, "birth").assertions,
    ...parseHijriYearPhrase(row.death_original_phrase, "death").assertions,
  ];
}

File: backend/app/hijri_date_phrase.py
"""Wortgleicher Python-Port von worker/src/core/hijri-date-phrase.mjs.

Zerlegt eine hidschri-Datumsangabe in ihre EINZELNEN belegten Moeglichkeiten.
Die Quellen nennen fuer eine Person regelmaessig mehrere Jahre nebeneinander
(«145، أو: 146هـ، أو: 147هـ، وقيل: 144هـ» sind vier Angaben, nicht eine).

Diese Funktion rechnet nicht: sie liest die Jahre, die dastehen, und behaelt den
Wortlaut jeder einzelnen Angabe bei. «أو» ist eine alternative Lesung innerhalb
derselben Aussage, «وقيل» eine eigenstaendige Aussage einer anderen Autoritaet --
der Unterschied bleibt erhalten.

Aenderungen hier und in der .mjs-Fassung muessen gemeinsam erfolgen;
tests/hijri-date-parity.check.mjs prueft beide gegen dieselbe Fixture.
"""

from __future__ import annotations

import re
from typing import Any

ARABIC_INDIC = "٠١٢٣٤٥٦٧٨٩"

QUALIFIERS: tuple[tuple[str, str], ...] = (
    ("بعيد", "shortly_after"),
    ("قبيل", "shortly_before"),
    ("بعد", "after"),
    ("قبل", "before"),
    ("نحو", "circa"),
    ("حدود", "circa"),
    ("زهاء", "circa"),
)

_REPORTED = re.compile("ق(?:ي|ِي)ل")
_ALTERNATIVE = re.compile("أو|او")
_YEAR = re.compile(r"\d{1,4}")
_TAIL = re.compile(r"^\s*(?:هـ|هجرية|هج)?")
_WORDING_HEAD = re.compile(r"^[\s،؛,.:]*(?:هـ|هجرية|هج)?[\s،؛,.:]*")


def to_ascii_digits(value: str) -> str:
    return "".join(str(ARABIC_INDIC.index(char)) if char in ARABIC_INDIC else char for char in value)


def _relation_for(connector: str, is_first: bool) -> str:
    if is_first:
        return "primary"
    if _REPORTED.search(connector):
        return "reported"
    if _ALTERNATIVE.search(connector):
        return "alternative"
    return "additional"


def _qualifier_for(segment: str) -> str | None:
    for word, code in QUALIFIERS:
        if word in segment:
            return code
    return None


def parse_hijri_year_phrase(phrase: Any, kind: str) -> list[dict[str, Any]]:
    """Liest alle Jahresangaben einer Phrase als einzelne Aussagen."""
    source_phrase = phrase.strip() if isinstance(phrase, str) and phrase.strip() else None
    if not source_phrase:
        return []

    ascii_text = to_ascii_digits(source_phrase)
    assertions: list[dict[str, Any]] = []
    seen: set[str] = set()
    previous_end = 0

    for match in _YEAR.finditer(ascii_text):
        year = int(match.group(0))
        # Ein Jahr ueber 1100 AH ist in diesem Bestand keine Jahresangabe,
        # sondern eine Seiten- oder Bandzahl in derselben Zeile.
        if not year or year > 1100:
            previous_end = match.end()
            continue

        connector = ascii_text[previous_end:match.start()]
        is_first = not assertions
        relation = _relation_for(connector, is_first)
        qualifier = _qualifier_for(ascii_text[:match.start()] if is_first else connector)

        tail = _TAIL.match(ascii_text[match.end():])
        wording_start = match.start() if is_first else previous_end
        wording_end = match.end() + (len(tail.group(0)) if tail else 0)
        wording = _WORDING_HEAD.sub("", source_phrase[wording_start:wording_end]).strip()

        key = f"{year}|{relation}|{qualifier or ''}"
        if key not in seen:
            seen.add(key)
            # Feldbild von ApiRijalDateAssertion (lib/api-client.ts).
            assertions.append({
                "kind": kind,
                "verb": None,
                "qualifier": qualifier,
                "valueAh": year,
                "approximate": qualifier == "circa",
                "rawPhrase": wording or str(year),
                "textOffset": match.start(),
                "evidenceClass": "rijal_statement",
                "confidence": None,
                "reviewStatus": "machine_unreviewed",
                "relation": relation,
                "sourcePhrase": source_phrase,
            })
        previous_end = match.end()

    return assertions
